docs/API-DESIGN.md

```diff
@@ -0,0 +1,865 @@
+    *   `DELETE /snapshot/cluster/{clusterName}/{snapshotName}`
+    *   `GET /snapshot/{clusterName}/{host}`
+    *   `POST /snapshot/{clusterName}/{host}`
+    *   `DELETE /snapshot/{clusterName}/{host}/{snapshotName}`
+*   Node
+    *   `GET /node/clientRequestLatencies/{clusterName}/{host}`
+    *   `GET /node/compactions/{clusterName}/{host}`
+    *   `GET /node/dropped/{clusterName}/{host}`
+    *   `GET /node/streams/{clusterName}/{host}`
+    *   `GET /node/tokens/{clusterName}/{host}`
+    *   `GET /node/tpstats/{clusterName}/{host}`
+*   Diagnostic Events
+    *   `GET /diag_event/sse_listen/{id}`
+    *   `GET /diag_event/subscription`
+    *   `POST /diag_event/subscription`
+    *   `DELETE /diag_event/subscription/{id}`
+    *   `GET /diag_event/subscription/{id}`
+
+# Common principles for the revised API
+
+1. Language conventions
+    1. The client interface should be named `reaper.Client` instead of `reaper.ReaperClient` to avoid repetition of the word "Reaper"
+    2. Use pointers to structs in API methods (no passing struct parameters by value)
+    3. Use Google's `uuid.UUID` for fields of uuid type
+    4. Use `big.Int` for Java `BigInteger`
+    5. Use `time.Time` for JodaTime `DateTime` structs
+        1. Deserialization will require a custom JSON unmarshaller because `DateTime` fields are sent as millis since the Epoch
+        2. Possible solution: `UnixTimeMillis`:
+
+				type UnixTimeMillis time.Time
+				func (t *UnixTimeMillis) UnmarshalJSON(data []byte) error {...}
+        
+    6. Use `github.com/google/go-querystring` to simplify creation of query strings from structs for optional REST parameters.
+    7. Use string-based types to capture Java enums, e.g. 
+	    
+			   type RepairSegmentState string
+			   const (
+					RepairSegmentStateNotStarted = RepairSegmentState("NOT_STARTED")
+					RepairSegmentStateRunning    = RepairSegmentState("RUNNING")
+					RepairSegmentStateDone       = RepairSegmentState("DONE")
+					RepairSegmentStateStarted    = RepairSegmentState("STARTED")
+				)
+        
+    8. Create type `Intensity` as an alias to float64 to better document its properties
+2. REST endpoint mappings
+    1. All API methods should take a first `context.Context `parameter, e.g. 
+
+				Ping(ctx context.Context) error
+        
+    2. Mandatory parameters should be captured in corresponding API method parameters, e.g. 
+
+				Login(ctx context.Context, username string, password string) error
+        
+    3. Optional parameters should be captured in dedicated structs, and the struct should be included as a last, nilable parameter, e.g. 
+
+				GetClusterNames(ctx context.Context, searchOptions *ClusterSearchOptions) ([]string, error)
+        
+    4. API methods should start with a standard prefix, followed by the REST resource name (`Cluster`, `RepairRun`, etc.):
+        1. GET: `Get` + plural when returning lists of entities, or singular when returning a single entity
```

Comment:
For the record, @burmanm suggested that we remove the `Get` prefix.
I also suggested alternative prefixes:
* `Find` for single-entity operations;
* `List` for multi-entity operations.

Reply:
I think the way [Effective Go recommends it](https://golang.org/doc/effective_go#Getters), we shouldn't put a prefix at all.
The team agreed to use Effective Go for its code style, and as weird as it is when you have a Java background (which is my case), I guess we should remove `Get` and not replace it with another prefix.

Reply:
Good point. Let's remove the getter prefix then. I will amend here and the other PR.